### PR TITLE
ONNXRT spec compilation

### PR DESCRIPTION
**Pull request by vancemiller**
This PR attempts to build ONNXRT in the ava-cuda-10.1 container

### CODE REVIEW

**Comment by vancemiller** [comment]
Currently failing with a linker error, missing dependency in the makefile?
```
[ 47%] Linking CXX executable worker
CMakeFiles/worker.dir/onnx_opt_nw_worker.c.o: In function `__helper_guestlib_init_prologue':
onnx_opt_nw_worker.c:(.text+0x320f): undefined reference to `guestlib_tf_opt_init'
CMakeFiles/worker.dir/onnx_opt_nw_worker.c.o: In function `__helper_guestlib_fini_prologue':
onnx_opt_nw_worker.c:(.text+0x3234): undefined reference to `guestlib_tf_opt_fini'
collect2: error: ld returned 1 exit status
```

**Comment by vancemiller** [comment]
steps to reproduce on this branch
```
pushd tools/docker
make ava-cuda-10.1
popd
mkdir build
./tools/docker/run_docker.sh ava-cuda-10.1
# now running commands in docker container
pushd build
cmake .. -DAVA_GEN_ONNXRT_SPEC=On -DAVA_MANAGER_LEGACY=On
make

File: cava/samples/onnxruntime/onnx_dump.c [hunk 1]
```diff
@@ -5,6 +5,7 @@ ava_number(10);
 ava_cflags(-I/usr/local/cuda-10.1/include -I../headers);
 ava_libs(-L/usr/local/cuda-10.1/lib64 -lcudart -lcuda -lcublas -lcudnn -lcufft -lcurand -lcusparse -lcusolver);
 ava_export_qualifier();
+ava_soname(cudart);
```
Comment (yuhc) [inline cava/samples/onnxruntime/onnx_dump.c:8]: I'm not leaning to specify soname for onnxrt and tf guestlib. As shown in line 6, they are collections of annotations for `-lcudart -lcuda -lcublas -lcudnn -lcufft -lcurand -lcusparse -lcusolver`.

Comment (yuhc) [reply cava/samples/onnxruntime/onnx_dump.c:8]: PS: Changing soname would require updating the spec's CMakeLists.txt as well.

**Review by yuhc — CHANGES_REQUESTED** [review]

File: cava/samples/onnxruntime/onnx_dump.c (hunk 1 above)
Comment (vancemiller) [reply cava/samples/onnxruntime/onnx_dump.c:8]: ah, it already generates libcudart.so anyways so this isn't necessary like it was for cudadrv

**Comment by vancemiller** [comment]
> at runtime there's a undefined symbol
> /data/NVIDIA_CUDA-10.1_Samples/0_Simple/vectorAdd/vectorAdd: symbol lookup error: /home/vance/Documents/workspace/a100-bench/build/ava/install/onnx_opt/lib/libcudart.so: undefined symbol: fatbin_handle_list
> 
> Looks like this is defined in worker/extensions/tf_optimization.c, should this be part of the guestlib compilation too?
> Hangchen — Today at 2:17 PM
> Yea we have to define it in guestlib too. It's referenced in an utility function but only used by worker. The utility functions are compiled into both guestlib and worker
> Ultimately we should avoid such dirty fix by distinguishing worker's and guestlib's utilities

How do we define this in the guestlib? What source file should this go in?

**Comment by yuhc** [comment]
Check ed73665b

**Comment by vancemiller** [comment]
Should the results of compiling onnx_dump be available on the client?

```
open /cuda_dumps/fatbin-info.ava [errno=2, errstr=No such file or directory] at /home/vance/Documents/workspace/a100-bench/third_party/ava/cava/samples/onnxruntime/onnx_opt.c:24701
```

**Comment by vancemiller** [comment]
got past it temporarily by disabling the preloaded cubins:
```
diff --git a/cava/samples/onnxruntime/onnx_opt.c b/cava/samples/onnxruntime/onnx_opt.c
index fa37310..943ec45 100644
--- a/cava/samples/onnxruntime/onnx_opt.c
+++ b/cava/samples/onnxruntime/onnx_opt.c
@@ -2,7 +2,7 @@ ava_name("CUDA Runtime for ONNX");
 ava_version("10.1.0");
 ava_identifier(ONNX_OPT);
 ava_number(10);
-ava_cflags(-I/usr/local/cuda-10.1/include -I../headers -DAVA_PRELOAD_CUBIN);
+ava_cflags(-I/usr/local/cuda-10.1/include -I../headers);
 ava_libs(-L/usr/local/cuda-10.1/lib64 -lcudart -lcuda -lcublas -lcudnn -lcufft -lcurand -lcusparse -lcusolver);
 ava_guestlib_srcs(extensions/tf_optimization.c extensions/cmd_batching.c);
 ava_worker_srcs(extensions/tf_optimization.c extensions/cmd_batching.c);
 ```

**Comment by vancemiller** [comment]
Now failing at runtime with no useful information
```
During startup program terminated with signal SIGSEGV, Segmentation fault.
```

**Comment by yuhc** [comment]
> Should the results of compiling onnx_dump be available on the client?
> 
> ```
> open /cuda_dumps/fatbin-info.ava [errno=2, errstr=No such file or directory] at /home/vance/Documents/workspace/a100-bench/third_party/ava/cava/samples/onnxruntime/onnx_opt.c:24701
> ```

Dumps must be present on both sides.
I can't really remember whether the opt specs work with AVA_PRELOAD_CUBIN disabled (I think it doesn't).

**Comment by vancemiller** [comment]
> 
> 
> > Should the results of compiling onnx_dump be available on the client?
> > ```
> > open /cuda_dumps/fatbin-info.ava [errno=2, errstr=No such file or directory] at /home/vance/Documents/workspace/a100-bench/third_party/ava/cava/samples/onnxruntime/onnx_opt.c:24701
> > ```
> 
> Dumps must be present on both sides.
> I can't really remember whether the opt specs work with AVA_PRELOAD_CUBIN disabled (I think it doesn't).

so there's some part of the build that is missing or not documented? What generates fatbin-info.ava and what installs it onto the client?

**Comment by vancemiller** [comment]
> ```
> Connect target API server (0.0.0.0:4005) at 0.0.0.0:4005
> Registering API command handler for API id 0: handler at 0x7f536532cee9
> 
> (process:1774893): GLib-CRITICAL **: 13:37:19.137: g_hash_table_lookup: assertion 'hash_table != NULL' failed
> deviceName is _Z9vectorAddPKfS0_Pfi
> vectorAdd: /home/vance/Documents/workspace/a100-bench/third_party/ava/cava/samples/onnxruntime/onnx_opt.c:976: __helper_assosiate_function: Assertion `0 && "func_id should not be null"' failed.
> ```
> 
> Maybe this is working as intended? I want to run kernels that aren't precompiled in this test but if I had an ONNX runtime benchmark this might work as is.

It seems like even if the hash table were available the code would still crash at onnx_opt.c:976 when running an arbitrary application because that `func_id` would not be found in the hash table generated from the `fatbin-info.ava` file. Does the hash table include other kernels besides the precompiled ones? Can I run arbitrary Cuda applications with this spec?

**Comment by yuhc** [comment]
> so there's some part of the build that is missing or not documented? What generates fatbin-info.ava and what installs it onto the client?

It's documented here `cava/samples/tensorflow/README.md`. I can send you a more detailed version if this is still unclear.

> It seems like even if the hash table were available the code would still crash at onnx_opt.c:976 when running an arbitrary application because that func_id would not be found in the hash table generated from the fatbin-info.ava file. Does the hash table include other kernels besides the precompiled ones? 

Do you have all `*.ava` files copied to `/cuda_dumps`?

> Can I run arbitrary Cuda applications with this spec?

Arbitrary CUDA applications can be run with the dump spec (you can remove the code that dumps CUDA binaries). Opt spec must use the dumps generated by the dump spec.

**Comment by vancemiller** [comment]
I can't run an application with the onnx_dump spec either as in the tensorflow README.  The application crashes after executing  `onnx_dump.c:285` when trying to open a dumped fatbinary on the client. The code seems to only dump fatbins on the worker. Is this a bug or am I doing something wrong?

```
    /*  Open the command pipe for reading */
    char pip_command[80];
    sprintf(pip_command, "/usr/local/cuda-10.1/bin/cuobjdump -elf /tmp/fatbin-%d.ava",
            ava_metadata(NULL)->num_fatbins);
    fp_pipe = popen(pip_command, "r");
    assert(fp_pipe);
```

**Comment by yuhc** [comment]
Whoops, I remembered it wrong. For the dump spec, guestlib and worker must be run on the same host (because only worker is dumping fatbins, but both need to read the fatbins). 
At the beginning both guestlib and worker are dumping fatbins and they can run on different hosts. But there were some race conditions in writing to the same file, so I made the above change.

**Comment by yuhc** [comment]
Tested with TF benchmarks. Both Release and Debug mode look good.

**Pull request merged**